src/jdk.hotspot.agent/share/classes/sun/jvm/hotspot/CommandProcessor.java

```diff
@@ -1794,32 +1794,31 @@ public void doit(Tokens t) {
                         /* First argument is compression level, second is filename */
                         /* Parse "gz=" option. */
                         gzlevel = parseHeapDumpCompressionLevel(option);
-                        if (gzlevel < 1) {
+                        if (gzlevel == 0) {
                             usage();
                             return;
                         }
                         /* Parse filename. */
                         filename = t.nextToken();
                     } else if (cntTokens == 1) {
                         filename = option;
-                        // Try to parse "gz=" option. If failed, treat it as filename
+                        // Try to parse "gz=" option.
                         if (option.startsWith("gz=")) {
                             gzlevel = parseHeapDumpCompressionLevel(option);
-                            if (gzlevel < 1) {
-                                err.println("Can not parse compression level from option \"" + option + "\".");
-                                if (gzlevel == 0) {
-                                    // Compression level not in range.
-                                    usage();
-                                    return;
-                                } else {
-                                    // Use the whole "gz=xxx" as dumped file name.
-                                    out.println("Use \"" + option +"\" as dumped file name.");
-                                }
-                            } else {
-                                filename = "heap.bin.gz";
+                            if (gzlevel == 0) {
+                                 usage();
+                                 return;
                             }
+                            filename = "heap.bin.gz";
                         }
                     }
+                    // Don't accept filename that start with "gz=" to avoid case like
+                    // "dumpheap gz=1 gz=2"
+                    if (filename.startsWith("gz=")) {
+                        err.println("Illegal filename \" + filename + \" should not start with \"gz=\".");
+                        usage();
+                        return;
+                    }
```

Comment:
Isn't this check only needed at the end of the `cntTokens == 2`  section? If there is just one token and it starts with `gz=`, it will be caught by `parseHeapDumpCompressionLevel`.

Reply:
> You are right.
> So, in my suggestion just replace this block:
> 
> ```
>                         if (gzlevel <= 0 || gzlevel > 9) {
>                             err.println("Invalid \"gz=\" option: " + option);
>                             usage();
>                             return;
>                         }
> ```
> 
> with:
> 
> ```
>                         if (gzlevel == 0) {
>                             usage();
>                             return;
>                         }
> ```

Thanks, just found there can be two prints of same error for option like gz=abc, 
I will change back:)